### PR TITLE
Food radius, regeneration and coordinates

### LINKED ISSUE #259: [FD] When a food is eaten its timer increases of one every tick
The following test should pass:
in food.cpp

```c++
#ifdef FIX_ISSUE_259
  {
    game g; //by default one uneaten food
    assert(has_food(g));
    auto initial_value_timer = get_nth_food_timer(g, 0);
    eat_nth_food(g, 0);
    assert(!has_food(g));
    g.tick();
    assert(init_value_timer + 1  == get_nth_food_timer(g, 0));
  }
#endif
```

### LINKED ISSUE #389: [FD] A food item has a radius
So far `food` objects are a single 1x1 square on the screen, but are drawn as if having a radius. Let's make a proper radius variable! 

```cpp
// A food has a radius member
  {
    const food f;
    assert(f.get_radius() >= 0.0);
  }
```

### PR DESCRIPTION
**Pull request by TheoPannetier**
Fix #389, #392 + removed food sprite from screen once eaten
Fix #259, #255, #394, now food items regen after 100 ticks 😁

### CODE REVIEW

File: game.cpp [hunk 1]
```diff
@@ -337,9 +337,22 @@ void game::tick()
   //Check and resolve wall collisions
   do_wall_collisions();
 
+  // Increment timers of all food elements
+  for (food &f : m_food)
+    {
+      f.increment_timer();
+    }
+
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done!

File: game.cpp [hunk 2]
```diff
@@ -539,12 +552,18 @@ void put_player_on_food(player &p, const food &f)
   p.place_to_position(get_position(f));
 }
 
+void put_player_near_food(player &p, const food &f, const double distance)
+{
+  std::vector<double> food_position = get_position(f);
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done!

File: game.cpp [hunk 3]
```diff
@@ -539,12 +552,18 @@ void put_player_on_food(player &p, const food &f)
...
+  food_position[0] += distance;
+  p.place_to_position(food_position);
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done !

File: game.cpp [hunk 4]
```diff
@@ -337,9 +337,22 @@ void game::tick()
...
   // Make players eat food
   make_players_eat_food();
 
+  // Regenerate food items
+  for (food &f : m_food)
+    {
+      if (f.is_eaten() && f.get_timer() >= f.get_regeneration_time())
+      f.set_food_state(food_state::uneaten);
+    }
+
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done!

File: game.cpp [hunk 5]
```diff
@@ -1399,6 +1457,34 @@ void test_game() //!OCLINT tests may be many
...
+    // Get player away so it does not eat food again
+    put_player_near_food(g.get_player(0), f, f.get_radius() * 2.0);
+
+    // Food item should not regen before the regeneration time
+    for(int i = 0; i != f_regen_time - 1; i++)
+      {
+        g.tick();
+      }
+    assert(nth_food_is_eaten(g,0));
+    g.tick();
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done!

File: game.cpp [hunk 6]
```diff
@@ -539,12 +552,18 @@ void put_player_on_food(player &p, const food &f)
...
+  food_position[0] += distance;
+  p.place_to_position(food_position);
+}
+
 bool have_same_position(const player& p, const food& f)
 {
-  return p.get_x() - f.get_x() < 0.0001 &&
-      p.get_x() - f.get_x() > -0.0001 &&
-      p.get_y() - f.get_y() < 0.0001 &&
-      p.get_y() - f.get_y() > -0.0001;
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done!

File: game.cpp [hunk 7]
```diff
@@ -539,12 +552,18 @@ void put_player_on_food(player &p, const food &f)
...
+  double f_radius = f.get_radius();
+  return abs(p.get_x() - f.get_x()) < f_radius &&
+      abs(p.get_y() - f.get_y()) < f_radius;
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp]: Done, I have overloaded `are_colliding()` for food vs players

File: game.cpp [hunk 8]
```diff
@@ -533,23 +539,41 @@ void shrink_losing_player(game &g)
   losing_player.shrink();
 }
 
-
 void put_player_on_food(player &p, const food &f)
 {
-  p.place_to_position(get_position(f));
+  coordinate f_p = f.get_position();
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp:544]:
```suggestion
  auto f_p = f.get_position();
```

File: game.cpp [hunk 9]
```diff
@@ -1269,6 +1326,30 @@ void test_game() //!OCLINT tests may be many
...
 
+#define FIX_ISSUE_392
+#ifdef FIX_ISSUE_392
+  //When a player gets within the radius of food it eats it
+  {
+    game g;
+    food f = g.get_food()[0];
+    double food_radius = f.get_radius();
+    double player_radius = get_nth_player_size(g, 0) / 2;
+    double collision_distance = food_radius + player_radius;
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp:1337]:
```suggestion
    double no_collision_distance = food_radius + player_radius;
```

File: game.cpp [hunk 10]
```diff
@@ -1485,7 +1596,7 @@ void test_game() //!OCLINT tests may be many
 #ifdef FIX_ISSUE_321
   {
     coordinate Some_random_point(1,1);
-    food n_food;
+    food n_;
```
[the remark this answers is not in the document]
Comment (TheoPannetier) [reply game.cpp:1599]:
```suggestion
    food n_food;
```

File: game.cpp [hunk 11]
```diff
@@ -1269,6 +1326,30 @@ void test_game() //!OCLINT tests may be many
...
+  //When a player gets within the radius of food it eats it
+  {
+    game g;
+    food f = g.get_food()[0];
+    double food_radius = f.get_radius();
+    double player_radius = get_nth_player_size(g, 0) / 2;
+    double no_collision_distance = food_radius + player_radius;
+
+    // Player at food radius should not trigger collision
+    put_player_near_food(g.get_player(0), f, collision_distance);
```
Comment (TheoPannetier) [inline game.cpp:1340]:
```suggestion
    put_player_near_food(g.get_player(0), f, no_collision_distance);
```

File: game.cpp [hunk 12]
```diff
@@ -1399,6 +1481,51 @@ void test_game() //!OCLINT tests may be many
...
+  {
+    game g;
+    double max_x = get_max_x(g);
+    assert(max_x == g.get_env().get_max_x());
+    double min_x = get_min_x(g);
+    assert(min_x == g.get_env().get_min_x());
+    double max_y = get_max_y(g);
+    assert(max_y == g.get_env().get_max_y());
+    double min_y = get_min_y(g);
+    assert(min_y == g.get_env().get_min_y());
```
Comment (swom) [inline game.cpp:1523]: wrapping at its best!

File: game.cpp [hunk 13]
```diff
@@ -1418,14 +1545,20 @@ void test_game() //!OCLINT tests may be many
     auto mean_x = calc_mean(food_x);
     auto mean_y = calc_mean(food_y);
 
-    assert(mean_x > -0.01 && mean_x < 0.01);
-    assert(mean_y > -0.01 && mean_y < 0.01);
+    // Mean position of food items should be center of game environment
+    double expected_mean_x = (get_max_x(g) + get_min_x(g)) / 2.0;
+    double expected_mean_y = (get_max_y(g) + get_min_y(g)) / 2.0;
```
Comment (swom) [inline game.cpp:1550]: shouldn't this be `max -min`?

**Review by swom — COMMENTED** [review]
I think you mixed a `+` with a `-`

File: game.cpp (hunk 13 above)
Comment (TheoPannetier) [reply game.cpp:1550]: No?

**Pull request merged**